src/sync.rs

```diff
@@ -487,25 +488,41 @@ impl<T> Default for FrozenVec<T> {
     }
 }
 
+const fn buffer_size(idx: usize) -> usize {
```

Comment:
nit: please explain this bitmath

Reply:
This looks to be copied from my prototype here: https://gist.github.com/Diggsey/c6a6ac87cf6516b05b5aaee183048177

The context for these functions is that we want to have a series of exponentially increasing buffer sizes. We want to maximize the total size of the buffers (since this determines the maximum size of the container) whilst minimizing the number of buffers (since we pay an up-front cost in space proportional to the number of buffers) without increasing the buffer size *too much* each time as this determines how much space will be wasted on average in allocated buffers. Finally, we also want a sequence will generates nice round numbers and is easy to work with.

`buffer_size`:
This is a compromise of the above factors: we multiple the buffer size by 4 each time whilst sizing the first buffer to 3, so the buffer sizes generated by the function will be [3, 12, 48, 192, etc.](https://oeis.org/A164346)

`prior_total_buffer_size`:
This computes the sum of the sizes of buffers prior to a particular buffer index, [aka 4^idx - 1](https://www.wolframalpha.com/input?i=sum+from+0+to+m+of+3+*+4%5En). The choice of sequence means that the total buffer size will always be a sequence of `1s` in binary, since it's a power of 2 minus one.

`buffer_index`
This determines which buffer contains the nth item (assuming the items are arranged sequentially in the buffers). Since the total buffer sizes are always sequences of 1s in binary, we can just could the number of binary digits in (i+1) and divide by 2 (rounding up). (That's what the `(65 - (i + 1).leading_zeros()) >> 1` part does. We use 65 rather than 64 so that the division by 2 rounds up instead of down. We divide by 2 (`>> 1`) because we skip every other power 2 since we increase the buffer size by 4 each time, and finally we subtract one because buffer indices are zero-indexed.

To give a worked example, the 3rd item has index 0b10 in binary:
- Add one: 0b11
- Count digits+1: 3
- Divide by 2: 1
- Subtract one: 0 (correct, this should be the last item in the first buffer)

For the 4th item, 0b11:
- Add one: 0b100
- Count digits+1: 4
- Divide by 2: 2
- Subtract one: 1 (correct, this should be the first item in the second buffer)

The `65` here should not be fixed, it should be the number of bits in a `usize` + 1.

Reply:
(oh, to be clear I had kinda reverse engineered the math here, but I was more asking for comments explaining what was going on because I'm not going to remember)